### PR TITLE
Check State Hashes while Processing Incoming Blocks

### PR DESCRIPTION
**Pull request by terenc3t**
To summarize what I added in this PR:

- Use `hash.Hash` instead of `common.Hash` for state fields so it integrates with black2b nicely
- Took out crosslink related fields in state object
- Wrote `hashActiveState` and `hashCrystallizedState` functions, they serialize state object, and hash it with blake2b
- Checked for active and crystallized hashes are correct in `canProcessBlock`
- Wrote negative test cases to test `canProcessBlock` when state hashes don't match

### CODE REVIEW

File: beacon-chain/blockchain/core.go [hunk 1]
```diff
@@ -144,6 +165,34 @@ func (b *BeaconChain) computeNewActiveState(seed common.Hash) (*types.ActiveStat
...
 }
 
+// hashActiveState serializes the active state object then uses
+// blake2b to hash the serialized object.
+func hashActiveState(state types.ActiveState) (hash.Hash, error) {
+	serializedState, err := rlp.EncodeToBytes(state)
+	if err != nil {
+		return nil, err
+	}
+	hash, err := blake2b.New256(serializedState)
```
Comment (rauljordan) [inline beacon-chain/blockchain/core.go]: return blake2b.new256(serializedState)

File: beacon-chain/blockchain/core.go [hunk 2]
```diff
@@ -144,6 +165,34 @@ func (b *BeaconChain) computeNewActiveState(seed common.Hash) (*types.ActiveStat
...
+}
+
+// hashCrystallizedState serializes the crystallized state object
+// then uses blake2b to hash the serialized object.
+func hashCrystallizedState(state types.CrystallizedState) (hash.Hash, error) {
+	serializedState, err := rlp.EncodeToBytes(state)
+	if err != nil {
+		return nil, err
+	}
+	hash, err := blake2b.New256(serializedState)
```
Comment (rauljordan) [inline beacon-chain/blockchain/core.go]: +1 to comment above

File: beacon-chain/blockchain/core_test.go [hunk 3]
```diff
@@ -195,7 +195,21 @@ func TestCanProcessBlock(t *testing.T) {
 	if _, err := beaconChain.CanProcessBlock(&faultyFetcher{}, block); err == nil {
 		t.Errorf("Using a faulty fetcher should throw an error, received nil")
 	}
+	activeState := &types.ActiveState{TotalAttesterDeposits: 10000}
+	beaconChain.state.ActiveState = activeState
 
+	activeHash, err := hashActiveState(*activeState)
+	if err != nil {
+		t.Fatalf("Can not hash active state: %v", err)
```
Comment (rauljordan) [inline beacon-chain/blockchain/core_test.go]: Cannot

File: beacon-chain/blockchain/core.go [hunk 4]
```diff
@@ -107,6 +110,24 @@ func (b *BeaconChain) CanProcessBlock(fetcher powchain.POWBlockFetcher, block *t
...
+	}
+
+	if !bytes.Equal(block.Data().ActiveStateHash.Sum(nil), hash.Sum(nil)) {
+		return false, fmt.Errorf("Active state hash mismatched, wanted: %v, got: %v", hash.Sum(nil), block.Data().ActiveStateHash.Sum(nil))
+	}
+	hash, err = hashCrystallizedState(*b.CrystallizedState())
+	if err != nil {
+		return false, err
+	}
+	if !bytes.Equal(block.Data().CrystallizedStateHash.Sum(nil), hash.Sum(nil)) {
```
Comment (rawfalafel) [inline beacon-chain/blockchain/core.go]: This part is likely to change so we can leave this be, but keep in mind that at most one every 100 blocks will contain the crystallized state. Each block's crystallized state can be up to 100MB, depending on the number of validators, and only changes after an epoch transition.

Comment (terenc3t) [reply beacon-chain/blockchain/core.go]: That's no longer the case with the latest 2.1 spec. Crystallized state changes every epoch which is `slot_number // EPOCH_LENGTH > epoch_number`  I'll make sure to add a condition for that when 2.1 gets finalized

Comment (rawfalafel) [reply beacon-chain/blockchain/core.go]: Ah, nice, didn't know that :+1:

**Pull request merged**